demos/device-shadow-for-aws-iot-embedded-sdk/shadow_demo_main.c

```diff
@@ -0,0 +1,679 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file shadow_demo_main.c
+ *
+ * @brief Demo for showing how to use the Device Shadow library's API. This version
+ * of Device Shadow API provide macros and helper functions for assembling MQTT topics
+ * strings, and for determining whether an incoming MQTT message is related to the
+ * device shadow. The Device Shadow library does not depend on a MQTT library,
+ * therefore the code for MQTT connections are placed in another file (shadow_demo_helpers.c)
+ * to make it easy to read the code using Device Shadow library.
+ *
+ * This example assumes there is a powerOn state in the device shadow. It does the
+ * following operations:
+ * 1. Establish a MQTT connection by using the helper functions in shadow_demo_helpers.c.
+ * 2. Assemble strings for the MQTT topics of device shadow, by using macros defined by the Device Shadow library.
+ * 3. Subscribe to those MQTT topics by using helper functions in shadow_demo_helpers.c.
+ * 4. Publish a desired state of powerOn by using helper functions in shadow_demo_helpers.c.  That will cause
+ * a delta message to be sent to device.
+ * 5. Handle incoming MQTT messages in prvEventCallback, determine whether the message is related to the device
+ * shadow by using a function defined by the Device Shadow library (Shadow_MatchTopic). If the message is a
+ * device shadow delta message, set a flag for the main function to know, then the main function will publish
+ * a second message to update the reported state of powerOn.
+ * 6. Handle incoming message again in prvEventCallback. If the message is from update/accepted, verify that it
+ * has the same clientToken as previously published in the update message. That will mark the end of the demo.
+ */
+
+/* Standard includes. */
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+
+/* Demo Specific configs. */
+#include "shadow_demo_config.h"
+
+#include "aws_demo.h"
+
+/* Kernel includes. */
+#include "FreeRTOS.h"
+#include "task.h"
+
+/* SHADOW API header. */
+#include "shadow.h"
+
+/* JSON library includes. */
+#include "core_json.h"
+
+/* Include header for connection configurations. */
+#include "aws_clientcredential.h"
+
+/* Include header for client credentials. */
+#include "aws_clientcredential_keys.h"
+
+/* Include header for root CA certificates. */
+#include "iot_default_root_certificates.h"
```

Comment:
Where are these headers being used in this file?

Reply:
will remove them.